### PR TITLE
Better handling of splat func arguments (prevent v8 deoptimization)

### PR DESCRIPTION
**Pull request by summivox**
see #857, jashkenas/coffeescript#3274 for discussion

### CODE REVIEW

**Comment by vendethiel** [comment]
Do we have a test for cases such as `(a, ...[x]:b, c) ->`?

File: src/ast.ls [hunk 1]
```diff
@@ -1641,14 +1641,17 @@ class exports.Assign extends Node
         else if (ret or len > 1) and (not ID.test rite.to-string! or left.assigns rite.to-string!)
             cache = sn(this, (rref = o.scope.temporary!), " = ", rite)
             rite  = rref
-        list = @"rend#{ left.constructor.display-name }" o, items, rite
+        if rite.to-string! is \arguments and not ret
+            arg = true
+            if left not instanceof Arr then @carp 'arguments can only destructure to array'
```
Comment (vendethiel) [inline src/ast.ls]: please newline that if

File: src/ast.ls [hunk 2]
```diff
@@ -1641,14 +1641,17 @@ class exports.Assign extends Node
         else if (ret or len > 1) and (not ID.test rite.to-string! or left.assigns rite.to-string!)
             cache = sn(this, (rref = o.scope.temporary!), " = ", rite)
             rite  = rref
-        list = @"rend#{ left.constructor.display-name }" o, items, rite
+        if rite.to-string! is \arguments and not ret
+            arg = true
```
Comment (vendethiel) [inline src/ast.ls]: i'd rather have a more meaningful name :)

File: src/ast.ls [hunk 3]
```diff
@@ -1930,7 +1946,7 @@ class exports.Fun extends Node
             # `(a = x) ->` => `(a ? x) ->`
             else if p.op is \=
                 params[i] = Binary (p.logic or \?), p.left, p.right
-        # `(a, ...b, c) ->` => `(a) -> [[] ...b, c] = @@`
+        # `(a, ...b, c) ->` => `(a) -> [[] ...b, c] = &`
```
Comment (vendethiel) [inline src/ast.ls]: good catch

File: src/ast.ls [hunk 4]
```diff
@@ -1939,30 +1955,28 @@ class exports.Fun extends Node
             params.0 = Var \it if body.traverse-children -> it.value is \it or null
         names   = []
         assigns = []
-        if params.length
-            dic = {}
```
Comment (vendethiel) [inline src/ast.ls]: was this variable unused?

File: src/ast.ls [hunk 5]
```diff
@@ -1939,30 +1955,28 @@ class exports.Fun extends Node
...
-                    assigns.push Assign vr, switch
-                        | df        => Binary p.op, v, p.second
-                        | has-unary => fold ((x, y) -> y.it = x; y), v, unaries.reverse!
-                        | otherwise => v
-                    vr = v
-                else if df
-                    assigns.push Assign vr, p.second, \=, p.op, true
-                names.push (scope.add vr.value, \arg, p), ', '
+        for p in params
+            vr = p
```
Comment (vendethiel) [inline src/ast.ls]: if you want to use this occasion to give more meaningful names to the variables around, feel very free to...

File: src/ast.ls (hunk 4 above)
Comment (summivox) [inline src/ast.ls]: I believe so. I did not find a reference in this function.

**Comment by summivox** [comment]
Ah you got me. A closure is generated in the case of `...[x]:b`.
I don't have time to fix it now though...

File: src/ast.ls (hunk 2 above)
Comment (summivox) [inline src/ast.ls]: better (refactorable) names on the way; I would rather not push before I fix the named nested splat though.

**Comment by vendethiel** [comment]
That's fine, don't rush it! Just wanted to make sure it worked.

**Comment by summivox** [comment]
I have a simple patch now but it is inelegant. It all boils down to this:
```livescript
[a]:b = [.. for til 4]
```
At top level this should not compile to a closure, but it does right now, i.e. #856 (funny I raised this one too).

**Comment by vendethiel** [comment]
> Do we have a test for cases such as (a, ...[x]:b, c) ->?

@gkz is that okay?

**Pull request merged**